src/details/ArborX_DetailsTreeConstruction.hpp

```diff
@@ -315,19 +311,40 @@ inline void TreeConstruction<DeviceType>::initializeLeafNodes(
   ARBORX_ASSERT(leaf_nodes.extent(0) == n);
 
   using Tag = typename Tag<decay_result_of_get_t<Access>>::type;
-  initializeLeafNodesDispatch(Tag{}, primitives, permutation_indices,
+  initializeLeafNodesDispatch(Tag{}, space, primitives, permutation_indices,
                               leaf_nodes);
 }
 
-template <typename DeviceType>
+template <typename ExecutionSpace, typename Primitives,
+          typename... PermutationIndicesViewProperties,
+          typename... LeafNodesViewProperties>
+inline void
+initializeLeafNodes(ExecutionSpace const &space, Primitives const &primitives,
+                    Kokkos::View<size_t *, PermutationIndicesViewProperties...>
+                        permutation_indices,
+                    Kokkos::View<Node *, LeafNodesViewProperties...> leaf_nodes)
+{
+  initializeLeafNodes(
+      space, primitives,
+      Kokkos::View<size_t const *, PermutationIndicesViewProperties...>{
+          permutation_indices},
+      leaf_nodes);
+}
+
+template <typename MemorySpace>
 class GenerateHierarchyFunctor
 {
 public:
+  template <typename... MortonCodesViewProperties,
+            typename... LeafNodesViewProperties,
+            typename... InternalNodesViewProperties,
+            typename... ParentsViewProperties>
   GenerateHierarchyFunctor(
-      Kokkos::View<unsigned int *, DeviceType> sorted_morton_codes,
-      Kokkos::View<Node *, DeviceType> leaf_nodes,
-      Kokkos::View<Node *, DeviceType> internal_nodes,
-      Kokkos::View<int *, DeviceType> parents)
+      Kokkos::View<unsigned int const *, MortonCodesViewProperties...>
+          sorted_morton_codes,
+      Kokkos::View<Node *, LeafNodesViewProperties...> leaf_nodes,
```

Comment:
If one of the view has a wrong `MemorySpace`, is the error easily understandable?

Reply:
Memory access violation at runtime.  We could add static assertions ion we really care.

Reply:
Why do you think runtime? Are views with different memory spaces assignable?

Reply:
Argument pack for the view properties will match anything.

Reply:
Right, but as you have `Kokkos::View<blah*, MemorySpace> _leaf_nodes` it will try to assign `Kokkos::View<blah*, LeafNodeViewProperties...>` to it. I am simply asking if you know what the error would look like in this case.